### PR TITLE
[flang][OpenMP] Support `target enter|update|exit .. nowait`

### PR DESCRIPTION
**Pull request by ergawy**
Extends `nowait` support for other device directives. This PR refactors the task generation utils used for the `target` directive so that they are general enough to be reused for other device directives as well.

### CODE REVIEW

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 1]
```diff
@@ -6403,16 +6401,45 @@ OpenMPIRBuilder::InsertPointTy OpenMPIRBuilder::createTargetData(
...
       assert(MapperFunc && "MapperFunc missing for standalone target data");
-      Builder.CreateCall(getOrCreateRuntimeFunctionPtr(*MapperFunc),
-                         OffloadingArgs);
+
+      auto TaskBodyCB = [&](Value *, Value *, IRBuilderBase::InsertPoint) {
+        if (Info.HasNoWait) {
+          OffloadingArgs.push_back(llvm::Constant::getNullValue(Int32));
+          OffloadingArgs.push_back(llvm::Constant::getNullValue(VoidPtr));
+          OffloadingArgs.push_back(llvm::Constant::getNullValue(Int32));
+          OffloadingArgs.push_back(llvm::Constant::getNullValue(VoidPtr));
```
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6419]:
```suggestion
          OffloadingArgs.append({
            llvm::Constant::getNullValue(Int32),
            llvm::Constant::getNullValue(VoidPtr),
            llvm::Constant::getNullValue(Int32),
            llvm::Constant::getNullValue(VoidPtr)
          });
```
(should work but did not test)

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 2]
```diff
@@ -6403,16 +6401,45 @@ OpenMPIRBuilder::InsertPointTy OpenMPIRBuilder::createTargetData(
...
+        }
+
+        Builder.CreateCall(getOrCreateRuntimeFunctionPtr(*MapperFunc),
+                           OffloadingArgs);
+
+        if (Info.HasNoWait) {
+          BasicBlock *OffloadContBlock =
+              BasicBlock::Create(Builder.getContext(), "omp_offload.cont");
+          auto *CurFn = Builder.GetInsertBlock()->getParent();
```
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6428]:
```suggestion
          BasicBlock *CurFn = Builder.GetInsertBlock()->getParent();
```
LLVM Coding Style applies

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 3]
```diff
@@ -7153,21 +7170,44 @@ emitTargetCall(OpenMPIRBuilder &OMPBuilder, IRBuilderBase &Builder,
...
-          /*DeviceID=*/nullptr, /*RTLoc=*/nullptr, AllocaIP, Dependencies,
-          HasNoWait));
+      Builder.restoreIP(OMPBuilder.emitTargetTask(TaskBodyCB,
+                                                  /*DeviceID=*/nullptr,
+                                                  /*RTLoc=*/nullptr, AllocaIP,
+                                                  Dependencies, HasNoWait));
     } else {
       Builder.restoreIP(EmitTargetCallFallbackCB(Builder.saveIP()));
     }
+
```
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:7210]:
```suggestion
```
[nit] unrelated change

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 4]
```diff
@@ -2905,9 +2907,9 @@ convertOmpTargetData(Operation *op, llvm::IRBuilderBase &builder,
             return success();
           })
           .Case([&](omp::TargetEnterDataOp enterDataOp) {
-            if (enterDataOp.getNowait())
+            if (!enterDataOp.getDependVars().empty())
```
Comment (Meinersbur) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2910]: Handling of `depend` looks unrelated to handling `nowait`. Since it reuses code, I don't mind in this case.

**Review by Meinersbur — APPROVED** [review]
LGTM

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 5]
```diff
@@ -6403,16 +6401,45 @@ OpenMPIRBuilder::InsertPointTy OpenMPIRBuilder::createTargetData(
...
+        Builder.CreateCall(getOrCreateRuntimeFunctionPtr(*MapperFunc),
+                           OffloadingArgs);
+
+        if (Info.HasNoWait) {
+          BasicBlock *OffloadContBlock =
+              BasicBlock::Create(Builder.getContext(), "omp_offload.cont");
+          auto *CurFn = Builder.GetInsertBlock()->getParent();
+          emitBranch(OffloadContBlock);
+          emitBlock(OffloadContBlock, CurFn, /*IsFinished=*/true);
```
[the remark this answers is not in the document]
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6430]: I think in this particular case it is not needed. Simply 

``
emitBlock(OffloadContBlock, CurFn, /*IsFinished=*/true);
``

 would do. But, I see this sequence of `emitBranch` and `emitBlock` used almost idiomatically in a number of places in `OMPIRBuilder.cpp`.

**Review by bhandarkar-pranav — APPROVED** [review]
A couple of minor nits, that's all.

**Review by NimishMishra — APPROVED** [review]
LGTM. Thanks for this.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 1 above)
Comment (ergawy) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp]: It does work. Thanks.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ergawy) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6431]: Seems like `emitBlock` is enough indeed. Let's see if the CI objects.

**Comment by ergawy** [comment]
Looks like this broke some buildbots. Looking into the reported failures ... 👀
